### PR TITLE
fix(prompt): cursor on prompt line, disallow ":edit"

### PR DESCRIPTION
**Pull request by shadmansaleh**
Working toward #34561

- [ ] ~~modified flag is not set upon editing prompt buffer~~
- [x] :edit is forbidden in prompt buffer
- [x] If the cursor is within the prompt text upon insert start jumps to end,

### CODE REVIEW

File: src/nvim/edit.c [hunk 1]
```diff
@@ -1532,7 +1532,9 @@ static void init_prompt(int cmdchar_todo)
 {
   char *prompt = prompt_text();
 
-  if (curwin->w_cursor.lnum < curbuf->b_prompt_start.mark.lnum) {
+  if (curwin->w_cursor.lnum < curbuf->b_prompt_start.mark.lnum
+      || (curwin->w_cursor.lnum == curbuf->b_prompt_start.mark.lnum
+          && (curwin->w_cursor.col < (int)strlen(prompt_text())))) {
```
Comment (shadmansaleh) [inline src/nvim/edit.c:1538]: Note (self): This breaks :normal O operation on prompt line.

File: src/nvim/ex_docmd.c [hunk 2]
```diff
@@ -5693,6 +5693,12 @@ static void ex_edit(exarg_T *eap)
     return;
   }
 
+  // prevent use of :edit on prompt-buffers
+  if (bt_prompt(curbuf) && eap->cmdidx == CMD_edit && *eap->arg == NUL) {
+    emsg(":edit on prompt buffer itself is not allowed");
```
Comment (justinmk) [inline src/nvim/ex_docmd.c]: Is there a way we can allow it but just don't empty the buffer? 
(I'm wondering if there are useful events that `:edit` triggers, which plugins might want.)

OTOH, we could try this and relax it later if needed.

```suggestion
    emsg("cannot :edit a prompt buffer");
```

Comment (shadmansaleh) [reply src/nvim/ex_docmd.c]: > Is there a way we can allow it but just don't empty the buffer?

What would :edit do exactly? If we leave as it is in master, `:edit` basically works as `clear history`. If we leave out the clearing part what is :edit actually doing? I kind of can't think of an usecase. Although to be honest you can only apply :edit on named buffers (I assume these buffers have a file associated with them) I thought with :edit it'd load the file content again. But on prompt buffer it doesn't do that it just clears it.

**Comment by shadmansaleh** [comment]
I have a very odd behavior that I can't seem to figure out. Basically if prompt has space in the end `prompt_setprompt(buf, "> ")`
Now if we use normal O operation in prompt line it pushes content of the prompt to line below and creates a new prompt in it's place with the prompt. 

Basically

```
> asdf|
```
becomes
```
> |
asdf
```
Now if I insert any text at all it works file. But if I don't insert anything and just `<esc>` to get out of insert mode the space in the prompt get's removed

```
>|
```

I don't get why this happens. I doesn't even happen if before you espace you add some text and remove them. Only immidiate esc seems to trigger this. Any clue what might be the reason?


Another thing is the tests are passing on my system when I'm running them locally even with a clean build. I'm confused why they are failing there..

I'm taking out the a on prompt text behavior change form this pr for above reasons

File: src/nvim/change.c [hunk 3]
```diff
@@ -154,7 +154,9 @@ void changed(buf_T *buf)
 /// Also used for recovery.
 void changed_internal(buf_T *buf)
 {
-  buf->b_changed = true;
+  if (!bt_prompt(buf)) {
+    buf->b_changed = true;
```
Comment (shadmansaleh) [inline src/nvim/change.c]: I have utterly no clue why this causes an unrelated test fail. It fails on this screen assert

https://github.com/neovim/neovim/blob/d9465e984b34ab7f98007e52edaf1ebc0ccef038/test/functional/legacy/prompt_buffer_spec.lua#L511-L521

The thing is it's not even an assert that's testing anything unique. It's just after `<cr>` / prompt-submit. And I can't reproduce the test failure on my local system. Only happening on CI.

Any clue why this might be happening? Or what can be done to fix it ?

Comment (justinmk) [reply src/nvim/change.c]: One does not merely set a buffer field :) Looking at assignments to `buf->b_changed`, I see things like this: https://github.com/neovim/neovim/blob/815916b45005da04838ceee749cec875ddad0f1a/src/nvim/change.c#L155-L158

That suggests that `changed_internal` may be the proper way to set b_changed.

Comment (shadmansaleh) [reply src/nvim/change.c]: It's still the same. I need to look at other ways to ignore modified status. The issue is that I can't reproduce the test failure locally.  Should I drop that commit and get the other two fixes merged?

Comment (justinmk) [reply src/nvim/change.c]: > should I drop that commit and get the other two fixes merged?

yeah let's deal with that in a separate PR

File: src/nvim/ex_docmd.c [hunk 4]
```diff
@@ -5693,6 +5693,12 @@ static void ex_edit(exarg_T *eap)
     return;
   }
 
+  // prevent use of :edit on prompt-buffers
+  if (bt_prompt(curbuf) && eap->cmdidx == CMD_edit && *eap->arg == NUL) {
+    emsg("cannot :edit a prompt buffer");
```
Comment (justinmk) [inline src/nvim/ex_docmd.c:5698]: need some sort of test coverage, just add 1 line that calls `:edit` in the "basic behavior" test , and assert with `pcall_err`.

File: src/nvim/edit.c [hunk 5]
```diff
@@ -1532,7 +1532,10 @@ static void init_prompt(int cmdchar_todo)
 {
   char *prompt = prompt_text();
 
-  if (curwin->w_cursor.lnum < curbuf->b_prompt_start.mark.lnum) {
+  if (curwin->w_cursor.lnum < curbuf->b_prompt_start.mark.lnum
+      || (cmdchar_todo != 'O'
+          && curwin->w_cursor.lnum == curbuf->b_prompt_start.mark.lnum
```
Comment (justinmk) [inline src/nvim/edit.c:1537]: is there a test we can update to try this behavior

Comment (shadmansaleh) [reply src/nvim/edit.c:1537]: We can add asserts for these cursor movements in multiline-input test . 
I'll add the test for the behavior changes.

Edit: tests added

File: src/nvim/edit.c [hunk 6]
```diff
@@ -1532,7 +1532,10 @@ static void init_prompt(int cmdchar_todo)
 {
   char *prompt = prompt_text();
 
-  if (curwin->w_cursor.lnum < curbuf->b_prompt_start.mark.lnum) {
+  if (curwin->w_cursor.lnum < curbuf->b_prompt_start.mark.lnum
+      || (cmdchar_todo != 'O'
```
Comment (justinmk) [inline src/nvim/edit.c:1536]: Not feeling great about special-casing things like this. But at least we have test coverage, so can iterate.

btw does lowercase `o` need to be handled?

File: test/functional/legacy/prompt_buffer_spec.lua [hunk 7]
```diff
@@ -130,6 +130,10 @@ describe('prompt buffer', function()
       {1:~                        }|*3
       {5:-- INSERT --}             |
     ]])
+    -- :edit doesn't apply on prompt buffer
+    local can_edit, _ = pcall(api.nvim_command, 'edit')
+    eq(can_edit, false, ':edit command in prompt buffer throws error')
```
Comment (justinmk) [inline test/functional/legacy/prompt_buffer_spec.lua]: `t.pcall_err`
```suggestion
    eq(':edit command in prompt buffer throws error',
       t.pcall_err(api.nvim_command, 'edit'))
```

File: src/nvim/edit.c (hunk 6 above)
Comment (shadmansaleh) [reply src/nvim/edit.c:1536]: No `o` doesn't need special handling here, because with `o` we just create a new line below current-prompt. `O` is basically an edge case, as in the case of `O`, we replace the current-prompt line.

File: test/functional/legacy/prompt_buffer_spec.lua (hunk 7 above)
Comment (shadmansaleh) [reply test/functional/legacy/prompt_buffer_spec.lua]: updated.
